File: test/language/choice_types/java/choice_types/UInt16ParamChoiceTest.java
package choice_types;

import static org.junit.jupiter.api.Assertions.*;

import java.io.IOException;

import org.junit.jupiter.api.Test;

import zserio.runtime.ZserioError;
import zserio.runtime.io.BitBuffer;
import zserio.runtime.io.ByteArrayBitStreamReader;
import zserio.runtime.io.ByteArrayBitStreamWriter;

import choice_types.uint16_param_choice.UInt16ParamChoice;

public class UInt16ParamChoiceTest
{
    @Test
    public void containerConstructor()
    {
        final int selector = VARIANT_A_SELECTOR;
        final UInt16ParamChoice uint16ParamChoice = new UInt16ParamChoice(selector);
        assertEquals(selector, uint16ParamChoice.getSelector());
    }

    @Test
    public void bitStreamReaderConstructor() throws IOException, ZserioError
    {
        final int selector = VARIANT_B_SELECTOR1;
        final int value = 234;
        final BitBuffer buffer = writeUInt16ParamChoiceToBitBuffer(selector, value);
        final ByteArrayBitStreamReader reader =
                new ByteArrayBitStreamReader(buffer.getBuffer(), buffer.getBitSize());
        final UInt16ParamChoice uint16ParamChoice = new UInt16ParamChoice(reader, selector);
        assertEquals(selector, uint16ParamChoice.getSelector());
        assertEquals((short)value, uint16ParamChoice.getValueB());
    }

    @Test
    public void choiceTag()
    {
        UInt16ParamChoice uint16ParamChoice = new UInt16ParamChoice(VARIANT_A_SELECTOR);
        assertEquals(UInt16ParamChoice.CHOICE_valueA, uint16ParamChoice.choiceTag());

        uint16ParamChoice = new UInt16ParamChoice(VARIANT_B_SELECTOR1);
        assertEquals(UInt16ParamChoice.CHOICE_valueB, uint16ParamChoice.choiceTag());

        uint16ParamChoice = new UInt16ParamChoice(VARIANT_C_SELECTOR);
        assertEquals(UInt16ParamChoice.CHOICE_valueC, uint16ParamChoice.choiceTag());

        uint16ParamChoice = new UInt16ParamChoice(EMPTY_SELECTOR1);
        assertEquals(UInt16ParamChoice.UNDEFINED_CHOICE, uint16ParamChoice.choiceTag());
    }

    @Test
    public void bitSizeOf()
    {
        final UInt16ParamChoice uint16ParamChoiceA = new UInt16ParamChoice(VARIANT_A_SELECTOR);
        final byte byteValueA = 99;
        uint16ParamChoiceA.setValueA(byteValueA);
        assertEquals(8, uint16ParamChoiceA.bitSizeOf());

        final UInt16ParamChoice uint16ParamChoiceB = new UInt16ParamChoice(VARIANT_B_SELECTOR2);
        final short shortValueB = 234;
        uint16ParamChoiceB.setValueB(shortValueB);
        assertEquals(16, uint16ParamChoiceB.bitSizeOf());

        final UInt16ParamChoice uint16ParamChoiceEmpty = new UInt16ParamChoice(EMPTY_SELECTOR1);
        assertEquals(0, uint16ParamChoiceEmpty.bitSizeOf());

        final UInt16ParamChoice uint16ParamChoiceC = new UInt16ParamChoice(VARIANT_C_SELECTOR);
        final int intValueC = 65535;
        uint16ParamChoiceC.setValueC(intValueC);
        assertEquals(32, uint16ParamChoiceC.bitSizeOf());
    }

    @Test
    public void getSelector()
    {
        final int selector = EMPTY_SELECTOR2;
        final UInt16ParamChoice uint16ParamChoice = new UInt16ParamChoice(selector);
        assertEquals(selector, uint16ParamChoice.getSelector());
    }

    @Test
    public void getSetA()
    {
        final UInt16ParamChoice uint16ParamChoice = new UInt16ParamChoice(VARIANT_A_SELECTOR);
        final byte value = 99;
        uint16ParamChoice.setValueA(value);
        assertEquals(value, uint16ParamChoice.getValueA());
    }

    @Test
    public void getSetB()
    {
        final UInt16ParamChoice uint16ParamChoice = new UInt16ParamChoice(VARIANT_B_SELECTOR3);
        final short value = 234;
        uint16ParamChoice.setValueB(value);
        assertEquals(value, uint16ParamChoice.getValueB());
    }

    @Test
    public void getSetC()
    {
        final UInt16ParamChoice uint16ParamChoice = new UInt16ParamChoice(VARIANT_C_SELECTOR);
        final int value = 65535;
        uint16ParamChoice.setValueC(value);
        assertEquals(value, uint16ParamChoice.getValueC());
    }

    @Test
    public void equals()
    {
        final UInt16ParamChoice uint16ParamChoice1 = new UInt16ParamChoice(VARIANT_A_SELECTOR);
        final UInt16ParamChoice uint16ParamChoice2 = new UInt16ParamChoice(VARIANT_A_SELECTOR);
        assertTrue(uint16ParamChoice1.equals(uint16ParamChoice2));

        final byte value = 99;
        uint16ParamChoice1.setValueA(value);
        assertFalse(uint16ParamChoice1.equals(uint16ParamChoice2));

        uint16ParamChoice2.setValueA(value);
        assertTrue(uint16ParamChoice1.equals(uint16ParamChoice2));

        final byte diffValue = value + 1;
        uint16ParamChoice2.setValueA(diffValue);
        assertFalse(uint16ParamChoice1.equals(uint16ParamChoice2));
    }

    @Test
    public void hashCodeMethod()
    {
        final UInt16ParamChoice uint16ParamChoice1 = new UInt16ParamChoice(VARIANT_A_SELECTOR);
        final UInt16ParamChoice uint16ParamChoice2 = new UInt16ParamChoice(VARIANT_A_SELECTOR);
        assertEquals(uint16ParamChoice1.hashCode(), uint16ParamChoice2.hashCode());

        final byte value = 99;
        uint16ParamChoice1.setValueA(value);
        assertTrue(uint16ParamChoice1.hashCode() != uint16ParamChoice2.hashCode());

        uint16ParamChoice2.setValueA(value);
        assertEquals(uint16ParamChoice1.hashCode(), uint16ParamChoice2.hashCode());

        final byte diffValue = value + 1;
        uint16ParamChoice2.setValueA(diffValue);
        assertTrue(uint16ParamChoice1.hashCode() != uint16ParamChoice2.hashCode());

        // use hardcoded values to check that the hash code is stable
        assertEquals(31623, uint16ParamChoice1.hashCode());
        assertEquals(31624, uint16ParamChoice2.hashCode());
    }

    @Test
    public void initializeOffsets()
    {
        final int bitPosition = 1;
        final UInt16ParamChoice uint16ParamChoiceA = new UInt16ParamChoice(VARIANT_A_SELECTOR);
        assertEquals(9, uint16ParamChoiceA.initializeOffsets(bitPosition));

        final UInt16ParamChoice uint16ParamChoiceB = new UInt16ParamChoice(VARIANT_B_SELECTOR2);
        assertEquals(17, uint16ParamChoiceB.initializeOffsets(bitPosition));

        final UInt16ParamChoice uint16ParamChoiceEmpty = new UInt16ParamChoice(EMPTY_SELECTOR1);
        assertEquals(1, uint16ParamChoiceEmpty.initializeOffsets(bitPosition));

        final UInt16ParamChoice uint16ParamChoiceC = new UInt16ParamChoice(VARIANT_C_SELECTOR);
        assertEquals(33, uint16ParamChoiceC.initializeOffsets(bitPosition));
    }

    @Test
    public void bitStreamWriterWrite() throws IOException, ZserioError
    {
        final UInt16ParamChoice uint16ParamChoiceA = new UInt16ParamChoice(VARIANT_A_SELECTOR);
        final byte byteValueA = 99;
        uint16ParamChoiceA.setValueA(byteValueA);
        ByteArrayBitStreamWriter writer = new ByteArrayBitStreamWriter();
        uint16ParamChoiceA.write(writer);
        ByteArrayBitStreamReader reader =
                new ByteArrayBitStreamReader(writer.toByteArray(), writer.getBitPosition());
        final UInt16ParamChoice readUInt16ParamChoiceA = new UInt16ParamChoice(reader, VARIANT_A_SELECTOR);
        assertEquals(byteValueA, readUInt16ParamChoiceA.getValueA());

        final UInt16ParamChoice uint16ParamChoiceB = new UInt16ParamChoice(VARIANT_B_SELECTOR1);
        final short shortValueB = 234;
        uint16ParamChoiceB.setValueB(shortValueB);
        writer = new ByteArrayBitStreamWriter();
        uint16ParamChoiceB.write(writer);
        reader = new ByteArrayBitStreamReader(writer.toByteArray(), writer.getBitPosition());
        final UInt16ParamChoice readUInt16ParamChoiceB = new UInt16ParamChoice(reader, VARIANT_B_SELECTOR1);
        assertEquals(shortValueB, readUInt16ParamChoiceB.getValueB());

        final UInt16ParamChoice uint16ParamChoiceC = new UInt16ParamChoice(VARIANT_C_SELECTOR);
        final int intValueC = 65535;
        uint16ParamChoiceC.setValueC(intValueC);
        writer = new ByteArrayBitStreamWriter();
        uint16ParamChoiceC.write(writer);
        reader = new ByteArrayBitStreamReader(writer.toByteArray(), writer.getBitPosition());
        final UInt16ParamChoice readUInt16ParamChoiceC = new UInt16ParamChoice(reader, VARIANT_C_SELECTOR);
        assertEquals(intValueC, readUInt16ParamChoiceC.getValueC());
    }

    private BitBuffer writeUInt16ParamChoiceToBitBuffer(int selector, int value) throws IOException
    {
        try (final ByteArrayBitStreamWriter writer = new ByteArrayBitStreamWriter())
        {
            switch (selector)
            {
            case 1:
                writer.writeByte((byte)value);
                break;

            case 2:
            case 3:
            case 4:
                writer.writeShort((short)value);
                break;

            case 5:
            case 6:
                break;

            default:
                writer.writeInt(value);
            }

            return new BitBuffer(writer.toByteArray(), writer.getBitPosition());
        }
    }

    private static int VARIANT_A_SELECTOR = 1;
    private static int VARIANT_B_SELECTOR1 = 2;
    private static int VARIANT_B_SELECTOR2 = 3;
    private static int VARIANT_B_SELECTOR3 = 4;
    private static int EMPTY_SELECTOR1 = 5;
    private static int EMPTY_SELECTOR2 = 6;
    private static int VARIANT_C_SELECTOR = 7;
}

File: test/extensions/language/choice_types/python/UInt16ParamChoiceTest.py
import zserio

import ChoiceTypes


class UInt16ParamChoiceTest(ChoiceTypes.TestCase):
    def testConstructor(self):
        selector = self.VARIANT_A_SELECTOR
        uint16ParamChoice = self.api.UInt16ParamChoice(selector)
        self.assertEqual(selector, uint16ParamChoice.selector)

        selector = self.VARIANT_B_SELECTOR1
        uint16ParamChoice = self.api.UInt16ParamChoice(selector, value_b_=1234)
        self.assertEqual(selector, uint16ParamChoice.selector)
        self.assertEqual(1234, uint16ParamChoice.value_b)

    def testFromReader(self):
        selector = self.VARIANT_B_SELECTOR1
        value = 234
        writer = zserio.BitStreamWriter()
        UInt16ParamChoiceTest._writeUInt16ParamChoiceToStream(writer, selector, value)
        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        uint16ParamChoice = self.api.UInt16ParamChoice.from_reader(reader, selector)
        self.assertEqual(selector, uint16ParamChoice.selector)
        self.assertEqual(value, uint16ParamChoice.value_b)

    def testEq(self):
        uint16ParamChoice1 = self.api.UInt16ParamChoice(self.VARIANT_A_SELECTOR)
        uint16ParamChoice2 = self.api.UInt16ParamChoice(self.VARIANT_A_SELECTOR)
        self.assertTrue(uint16ParamChoice1 == uint16ParamChoice2)

        value = 99
        uint16ParamChoice1.value_a = value
        self.assertFalse(uint16ParamChoice1 == uint16ParamChoice2)

        uint16ParamChoice2.value_a = value
        self.assertTrue(uint16ParamChoice1 == uint16ParamChoice2)

        diffValue = value + 1
        uint16ParamChoice2.value_a = diffValue
        self.assertFalse(uint16ParamChoice1 == uint16ParamChoice2)

    def testHash(self):
        uint16ParamChoice1 = self.api.UInt16ParamChoice(self.VARIANT_A_SELECTOR)
        uint16ParamChoice2 = self.api.UInt16ParamChoice(self.VARIANT_A_SELECTOR)
        self.assertEqual(hash(uint16ParamChoice1), hash(uint16ParamChoice2))

        value = 99
        uint16ParamChoice1.value_a = value
        self.assertTrue(hash(uint16ParamChoice1) != hash(uint16ParamChoice2))

        uint16ParamChoice2.value_a = value
        self.assertEqual(hash(uint16ParamChoice1), hash(uint16ParamChoice2))

        diffValue = value + 1
        uint16ParamChoice2.value_a = diffValue
        self.assertTrue(hash(uint16ParamChoice1) != hash(uint16ParamChoice2))

        # use hardcoded values to check that the hash code is stable
        # using __hash__ to prevent 32-bit Python hash() truncation
        self.assertEqual(31623, uint16ParamChoice1.__hash__())
        self.assertEqual(31624, uint16ParamChoice2.__hash__())

    def testGetSelector(self):
        selector = self.EMPTY_SELECTOR2
        uint16ParamChoice = self.api.UInt16ParamChoice(selector)
        self.assertEqual(selector, uint16ParamChoice.selector)

    def testGetSetA(self):
        uint16ParamChoice = self.api.UInt16ParamChoice(self.VARIANT_A_SELECTOR)
        value = 99
        uint16ParamChoice.value_a = value
        self.assertEqual(value, uint16ParamChoice.value_a)

    def testGetSetB(self):
        uint16ParamChoice = self.api.UInt16ParamChoice(self.VARIANT_B_SELECTOR3)
        value = 234
        uint16ParamChoice.value_b = value
        self.assertEqual(value, uint16ParamChoice.value_b)

    def testGetSetC(self):
        uint16ParamChoice = self.api.UInt16ParamChoice(self.VARIANT_C_SELECTOR)
        value = 65535
        uint16ParamChoice.value_c = value
        self.assertEqual(value, uint16ParamChoice.value_c)

    def testChoiceTag(self):
        uint16ParamChoice = self.api.UInt16ParamChoice(self.VARIANT_A_SELECTOR)
        self.assertEqual(uint16ParamChoice.CHOICE_VALUE_A, uint16ParamChoice.choice_tag)

        uint16ParamChoice = self.api.UInt16ParamChoice(self.VARIANT_B_SELECTOR1)
        self.assertEqual(uint16ParamChoice.CHOICE_VALUE_B, uint16ParamChoice.choice_tag)

        uint16ParamChoice = self.api.UInt16ParamChoice(self.VARIANT_C_SELECTOR)
        self.assertEqual(uint16ParamChoice.CHOICE_VALUE_C, uint16ParamChoice.choice_tag)

        uint16ParamChoice = self.api.UInt16ParamChoice(self.EMPTY_SELECTOR1)
        self.assertEqual(uint16ParamChoice.UNDEFINED_CHOICE, uint16ParamChoice.choice_tag)

    def testBitSizeOf(self):
        uint16ParamChoiceA = self.api.UInt16ParamChoice(self.VARIANT_A_SELECTOR)
        byteValueA = 99
        uint16ParamChoiceA.value_a = byteValueA
        self.assertEqual(8, uint16ParamChoiceA.bitsizeof())

        uint16ParamChoiceB = self.api.UInt16ParamChoice(self.VARIANT_B_SELECTOR2)
        shortValueB = 234
        uint16ParamChoiceB.value_b = shortValueB
        self.assertEqual(16, uint16ParamChoiceB.bitsizeof())

        uint16ParamChoiceEmpty = self.api.UInt16ParamChoice(self.EMPTY_SELECTOR1)
        self.assertEqual(0, uint16ParamChoiceEmpty.bitsizeof())

        uint16ParamChoiceC = self.api.UInt16ParamChoice(self.VARIANT_C_SELECTOR)
        intValueC = 65535
        uint16ParamChoiceC.value_c = intValueC
        self.assertEqual(32, uint16ParamChoiceC.bitsizeof())

    def testInitializeOffsets(self):
        bitPosition = 1
        uint16ParamChoiceA = self.api.UInt16ParamChoice(self.VARIANT_A_SELECTOR)
        self.assertEqual(9, uint16ParamChoiceA.initialize_offsets(bitPosition))

        uint16ParamChoiceB = self.api.UInt16ParamChoice(self.VARIANT_B_SELECTOR2)
        self.assertEqual(17, uint16ParamChoiceB.initialize_offsets(bitPosition))

        uint16ParamChoiceEmpty = self.api.UInt16ParamChoice(self.EMPTY_SELECTOR1)
        self.assertEqual(1, uint16ParamChoiceEmpty.initialize_offsets(bitPosition))

        uint16ParamChoiceC = self.api.UInt16ParamChoice(self.VARIANT_C_SELECTOR)
        self.assertEqual(33, uint16ParamChoiceC.initialize_offsets(bitPosition))

    def testReadWrite(self):
        uint16ParamChoiceA = self.api.UInt16ParamChoice(self.VARIANT_A_SELECTOR)
        byteValueA = 99
        uint16ParamChoiceA.value_a = byteValueA
        writer = zserio.BitStreamWriter()
        uint16ParamChoiceA.write(writer)
        readUInt16ParamChoiceA = self.api.UInt16ParamChoice(self.VARIANT_A_SELECTOR)
        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        readUInt16ParamChoiceA.read(reader)
        self.assertEqual(byteValueA, readUInt16ParamChoiceA.value_a)
        self.assertEqual(uint16ParamChoiceA, readUInt16ParamChoiceA)

        shortValueB = 234
        uint16ParamChoiceB = self.api.UInt16ParamChoice(self.VARIANT_B_SELECTOR1, value_b_=shortValueB)
        writer = zserio.BitStreamWriter()
        uint16ParamChoiceB.write(writer)
        readUInt16ParamChoiceB = self.api.UInt16ParamChoice(self.VARIANT_B_SELECTOR1)
        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        readUInt16ParamChoiceB.read(reader)
        self.assertEqual(shortValueB, readUInt16ParamChoiceB.value_b)
        self.assertEqual(uint16ParamChoiceB, readUInt16ParamChoiceB)

        uint16ParamChoiceC = self.api.UInt16ParamChoice(self.VARIANT_C_SELECTOR)
        intValueC = 65535
        uint16ParamChoiceC.value_c = intValueC
        writer = zserio.BitStreamWriter()
        uint16ParamChoiceC.write(writer)
        readUInt16ParamChoiceC = self.api.UInt16ParamChoice(self.VARIANT_C_SELECTOR)
        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        readUInt16ParamChoiceC.read(reader)
        self.assertEqual(intValueC, readUInt16ParamChoiceC.value_c)
        self.assertEqual(uint16ParamChoiceC, readUInt16ParamChoiceC)

    @staticmethod
    def _writeUInt16ParamChoiceToStream(writer, selector, value):
        if selector == 1:
            writer.write_signed_bits(value, 8)
        elif selector in (2, 3, 4):
            writer.write_signed_bits(value, 16)
        elif selector in (5, 6):
            pass
        else:
            writer.write_signed_bits(value, 32)

    VARIANT_A_SELECTOR = 1
    VARIANT_B_SELECTOR1 = 2
    VARIANT_B_SELECTOR2 = 3
    VARIANT_B_SELECTOR3 = 4
    EMPTY_SELECTOR1 = 5
    EMPTY_SELECTOR2 = 6
    VARIANT_C_SELECTOR = 7
